systemtest/produce_consume_basic_test.go

```diff
@@ -0,0 +1,172 @@
+				}),
+				consumer: newKafkaConsumer(t, kafka.ConsumerConfig{
+					Logger:  logger,
+					Decoder: json.JSON{},
+					Topics:  topics,
+					GroupID: t.Name(),
+					Processor: assertBatchFunc(t, consumerAssertions{
+						records:   &records,
+						processor: model.TransactionProcessor,
+					}),
+				}),
+				timeout: timeout,
+			})
+		})
+		t.Run("PubSubLite"+name, func(t *testing.T) {
+			err := ProvisionPubSubLite(context.Background(),
+				newPubSubLiteConfig(topics...),
+			)
+			require.NoError(t, err)
+
+			ctx, cancel := context.WithTimeout(context.Background(), timeout)
+			defer cancel()
+			var records atomic.Int64
+			testProduceConsume(ctx, t, produceConsumeCfg{
+				events:  events,
+				records: &records,
+				producer: newPubSubLiteProducer(ctx, t, pubsublite.ProducerConfig{
+					Topics:      []apmqueue.Topic{apmqueue.Topic(topics[0])},
+					Logger:      logger,
+					Encoder:     json.JSON{},
+					TopicRouter: topicRouter,
+					Sync:        sync,
+				}),
+				consumer: newPubSubLiteConsumer(ctx, t, pubsublite.ConsumerConfig{
+					Logger:        logger,
+					Decoder:       json.JSON{},
+					Subscriptions: pubSubLiteSubscriptions(topics...),
+					Processor: assertBatchFunc(t, consumerAssertions{
+						records:   &records,
+						processor: model.TransactionProcessor,
+					}),
+				}),
+				timeout: timeout,
+			})
+		})
+	}
+}
+
+type produceConsumeCfg struct {
+	events   int
+	producer apmqueue.Producer
+	consumer apmqueue.Consumer
+	records  *atomic.Int64
+	timeout  time.Duration
+}
+
+func testProduceConsume(ctx context.Context, t testing.TB, cfg produceConsumeCfg) {
+	// Run consumer and assert that the events are eventually set.
+	go cfg.consumer.Run(ctx)
+	batch := make(model.Batch, 0, cfg.events)
+	for i := 0; i < cfg.events; i++ {
+		batch = append(batch, model.APMEvent{
+			Timestamp: time.Now(),
+			Processor: model.TransactionProcessor,
+			Trace:     model.Trace{ID: fmt.Sprintf("trace-%d", i+1)},
+			Event: model.Event{
+				Duration: time.Millisecond * (time.Duration(rand.Int63n(999)) + 1),
+			},
+			Transaction: &model.Transaction{
+				ID: fmt.Sprintf("transaction-%d", i+1),
+			},
+		})
+	}
+
+	// Produce the records to queue.
+	assert.NoError(t, cfg.producer.ProcessBatch(ctx, &batch))
+	assert.Eventually(t, func() bool {
+		return cfg.records.Load() == int64(cfg.events) // Assertion
+	},
+		cfg.timeout,          // Timeout
+		100*time.Millisecond, // Poll
+		"expected records (%d) records do not match consumed records (%d)", // ErrMessage
+		cfg.events,
+		cfg.records.Load(),
+	)
+}
+
+type consumerAssertions struct {
+	processor model.Processor
+	records   *atomic.Int64
+}
+
+func assertBatchFunc(t testing.TB, assertions consumerAssertions) model.BatchProcessor {
+	return model.ProcessBatchFunc(func(_ context.Context, b *model.Batch) error {
+		assert.Greater(t, len(*b), 0)
+		for _, r := range *b {
+			assert.Equal(t, r.Processor, assertions.processor, r)
+			if assertions.records != nil {
```

Comment:
**question**: I might be missing something, is this ever nil ?

Reply:
not in this test, but other tests may choose to not pass in any `records` counter.

Reply:
Thanks for the clarification!